### PR TITLE
feat(alerting): Implement new Teams Workflow alert

### LINKED ISSUE #820: Change the Teams alert from the Office 365 connector to the Teams Workflow App Webhook Connector
### Describe the feature request

Recently, Teams alerts in my tenant from Gatus have had the following text appended to them:

"**Action Required:** O365 connectors within Teams will be deprecated and notifications from this service will stop. [Learn more](https://aka.ms/O365ConnectorDeprecation)
about the timing and how the Workflows app provides a more flexible and secure experience. If you want to continue receiving these types of messages, you can use a workflow to post messages from a webhook request.
[Set up workflow](https://teams.microsoft.com/l/task/c3a1996d-db0f-4857-a6ea-7aabf0266b00?url=https://make.powerautomate.com/teams/store/flowCreation?templateId=d271a6f01c2545a28348d8f2cddf4c8f)"

So, reading through the first link, all Office connectors, including the Incoming Webhook connector Gatus uses are being retired in favour of the Teams Workflow App which seems to be a client for Power Automate.

There appears to be webhook connector template in the new workflow app that will operate in a similar way to the O365 webhook connector.

This (https://learn.microsoft.com/en-us/connectors/teams/?tabs=text1#microsoft-teams-webhook) page describes JSON body content required, which is in the Adaptive Card JSON format.

### Why do you personally want this feature to be implemented?

I use as Teams my main conduit for important alerts.

### How long have you been using this project?

6 months

### Additional information

I'm curious to know if other users who have configured Teams alerts are also receiving this message?

This is great project, thank you for all the hard work!

### PR DESCRIPTION
**Pull request by JamesHillyard**
## Summary
Fixes #820 by introducing a new `teams-workflows` alert compatible with Teams Workflows. Microsoft is deprecating the O365 webhook connector for Teams Workflows with MS Power Automate ([Source: Microsoft DevBlog](https://devblogs.microsoft.com/microsoft365dev/retirement-of-office-365-connectors-within-microsoft-teams/)).

The new Workflows require [Microsoft Adaptive Cards](https://adaptivecards.io/) and have a proprietary JSON format that means the old `teams` connector using a new Workflow URL will fail with this error:
```
The execution of template action 'Send_each_adaptive_card' failed: the result of the evaluation of 'foreach' expression '@triggerOutputs()?['body']?['attachments']' is of type 'Null'. The result must be a valid array.
```

## Breaking Changes
**There are no breaking changes**. The old `teams` alert still exists and will work with legacy O365 Webhook URLs.

There are no breaking changes in configuring the `teams` and `teamswebhook` alerts, meaning migration is simple.

## Migrating to new Teams Webhooks
End-users will need to make the following changes to migrate from the legacy `teams` alerts to `teamswebhook` alerts:
- Create a Workflow in Microsoft Teams for incoming webhooks. [Microsoft Documentation](https://support.microsoft.com/en-us/office/create-incoming-webhooks-with-workflows-for-microsoft-teams-8ae491c7-0394-4861-ba59-055e33f75498)
- Replace all occurences of `teams` (In `alerting` and `endpoints[].alerts`) with `teams-workflows`
- Replace the `alerting.teams.webhook-url` (Now `alerting.teams-workflows.webhook-url`) with the new URL generated in the Teams Workflow

### MS Teams Configuration
image

### Gatus Configuration
```yml
alerting:
  teams-workflows:
    webhook-url: "[WORKFLOW-WEBHOOK-URL-HERE]"
    default-alert:
      description: ""
      send-on-resolved: true
      failure-threshold: 3
      success-threshold: 3


endpoints:
  # Website Group
  - name: front-end
    url: https://www.google.com
    interval: 3s
    conditions:
      - "[STATUS] == 200"
      - "[RESPONSE_TIME] < 50"
    alerts:
      - type: teams-workflows
        description: "Description Lies Here"
```

## Remaining Issues
There are a few remaining issues, none are blockers to using this feature, but likely should be addressed in the future with further investigation.

### Adaptive Card Version >1.5
Using an Adaptive Card higher than 1.4 (1.6 is latest) results in the message showing 
> We're sorry, this card couldn't be displayed

This has been reported to AdaptiveCards ([Report](https://www.github.com/microsoft/AdaptiveCards/issues/7676)), but discussion around this issue seems very low.

### Card Width
The cards do not show on teams as full-width. A [StackOverflow post](https://stackoverflow.com/questions/59728112/increase-the-width-of-adaptive-card-for-ms-teams-channel) would suggest the width can be changed, but I wasn't able to get this to work.

### Card Colour Strip
The red and green colour strip on the old teams cards at the top for triggered/resolved is not supported in Adaptive Cards. There are workaround options as identified in this [Stack Overflow post](https://stackoverflow.com/questions/72873499/adding-a-colored-strip-on-the-top-of-adaptive-card) but it doesn't look great.

## Checklist
- [X] Tested and/or added tests to validate that the changes work as intended, if applicable.
- [X] Updated documentation in `README.md`, if applicable.

### CODE REVIEW

File: alerting/config.go [hunk 1]
```diff
@@ -89,6 +90,9 @@ type Config struct {
 	// Teams is the configuration for the teams alerting provider
 	Teams *teams.AlertProvider `yaml:"teams,omitempty"`
 
+	// TeamsWorkflow is the configuration for the teams alerting provider using the new Workflow App Webhook Connector
+	TeamsWorkflow *teamsworkflow.AlertProvider `yaml:"teamsworkflow,omitempty"`
```
Comment (TwiN) [inline alerting/config.go:94]:
```suggestion
	TeamsWorkflow *teamsworkflow.AlertProvider `yaml:"teams-workflow,omitempty"`
```

File: README.md [hunk 2]
```diff
@@ -564,7 +565,8 @@ endpoints:
 | `alerting.pagerduty`      | Configuration for alerts of type `pagerduty`. <br />See [Configuring PagerDuty alerts](#configuring-pagerduty-alerts).                   | `{}`    |
 | `alerting.pushover`       | Configuration for alerts of type `pushover`. <br />See [Configuring Pushover alerts](#configuring-pushover-alerts).                      | `{}`    |
 | `alerting.slack`          | Configuration for alerts of type `slack`. <br />See [Configuring Slack alerts](#configuring-slack-alerts).                               | `{}`    |
-| `alerting.teams`          | Configuration for alerts of type `teams`. <br />See [Configuring Teams alerts](#configuring-teams-alerts).                               | `{}`    |
+| `alerting.teams`          | Configuration for alerts of type `teams`. *(Deprecated)* <br />See [Configuring Teams alerts](#configuring-teams-alerts-deprecated).     | `{}`    |
+| `alerting.teamsworkflow`  | Configuration for alerts of type `teamsworkflow`. <br />See [Configuring Teams Workflow alerts](#configuring-teams-workflow-alerts).     | `{}`    |
```
Comment (TwiN) [inline README.md:569]:
```suggestion
| `alerting.teamsworkflow`  | Configuration for alerts of type `teamsworkflow`. <br />See [Configuring Teams Workflow alerts](#configuring-teams-workflow-alerts).     | `{}`    |
```
could you rename `teamsworkflow` to `teams-workflows` here?

File: README.md [hunk 3]
```diff
@@ -1220,6 +1227,61 @@ Here's an example of what the notifications look like:
...
+| `alerting.teamsworkflow`                         | Configuration for alerts of type `teams`                                                   | `{}`               |
+| `alerting.teamsworkflow.webhook-url`             | Teams Webhook URL                                                                          | Required `""`      |
+| `alerting.teamsworkflow.default-alert`           | Default alert configuration. <br />See [Setting a default alert](#setting-a-default-alert) | N/A                |
+| `alerting.teamsworkflow.overrides`               | List of overrides that may be prioritized over the default configuration                   | `[]`               |
+| `alerting.teamsworkflow.title`                   | Title of the notification                                                                  | `"&#x26D1; Gatus"` |
+| `alerting.teamsworkflow.overrides[].group`       | Endpoint group for which the configuration will be overridden by this configuration        | `""`               |
+| `alerting.teamsworkflow.overrides[].webhook-url` | Teams WorkFlow Webhook URL                                                                 | `""`               |
```
Comment (TwiN) [inline README.md:1243]: Same here

File: README.md [hunk 4]
```diff
@@ -67,7 +67,8 @@ Have any feedback or questions? [Create a discussion](https://github.com/TwiN/ga
     - [Configuring PagerDuty alerts](#configuring-pagerduty-alerts)
     - [Configuring Pushover alerts](#configuring-pushover-alerts)
     - [Configuring Slack alerts](#configuring-slack-alerts)
-    - [Configuring Teams alerts](#configuring-teams-alerts)
+    - [Configuring Teams alerts *(Deprecated)*](#configuring-teams-alerts-deprecated)
+    - [Configuring Teams Workflow alerts](#configuring-teams-workflow-alerts)
```
Comment (TwiN) [inline README.md:71]:
```suggestion
    - [Configuring Teams Workflows alerts](#configuring-teams-workflows-alerts)
```

File: README.md [hunk 5]
```diff
@@ -1220,6 +1227,61 @@ Here's an example of what the notifications look like:
...
+    group: core
+    url: "https://example.org/"
+    interval: 5m
+    conditions:
+      - "[STATUS] == 200"
+      - "[CERTIFICATE_EXPIRATION] > 48h"
+    alerts:
+      - type: teamsworkflow
+        description: "healthcheck failed"
+        send-on-resolved: true
```
Comment (TwiN) [inline README.md:1278]:
```suggestion
alerting:
  teams-workflows:
    webhook-url: "https://********.webhook.office.com/webhookb2/************"
    # You can also add group-specific to keys, which will
    # override the to key above for the specified groups
    overrides:
      - group: "core"
        webhook-url: "https://********.webhook.office.com/webhookb3/************"

endpoints:
  - name: website
    url: "https://twin.sh/health"
    interval: 30s
    conditions:
      - "[STATUS] == 200"
      - "[BODY].status == UP"
      - "[RESPONSE_TIME] < 300"
    alerts:
      - type: teams-workflows
        description: "healthcheck failed"
        send-on-resolved: true
```

File: alerting/provider/teamsworkflow/teamsworkflow.go [hunk 6]
```diff
@@ -0,0 +1,182 @@
+package teamsworkflow
```
Comment (TwiN) [inline alerting/provider/teamsworkflow/teamsworkflow.go:1]:
```suggestion
package teamsworkflows
```
the path should also be changed.

**Review by TwiN — CHANGES_REQUESTED** [review]
Great work! The only real change I see here is renaming teamsworkflow` to `teams-workflows`.

Not a super big deal, but looking at https://support.microsoft.com/en-us/office/creating-a-workflow-from-a-channel-in-teams-242eb8f2-f328-45be-b81f-9817b51a5f0e, it seems like their service is "Workflows", not "Workflow", and well, I'd rather avoid confusion.

In a later version of Gatus, I'll probably get rid of the current `teams` once they get rid of the O365 connectors and replace it for this implementation.

**Comment by JamesHillyard** [comment]
@TwiN Makes sense, I have renamed `teamsworkflows` to `teams-workflows` and tested it still works. Thanks for the review!

**Review by TwiN — APPROVED** [review]

**Comment by TwiN** [comment]
@JamesHillyard Looks like some of your tests are failing

```
 --- FAIL: TestAlertProvider_buildRequestBody (0.00s)
    --- FAIL: TestAlertProvider_buildRequestBody/triggered (0.00s)
        teamsworkflows_test.go:189: expected:
            {"attachments":[{"content":{"type":"AdaptiveCard","version":"1.4","body":[{"type":"TextBlock","text":"\u0026#x1f6a8 Gatus","wrap":false,"size":"Medium","weight":"Bolder"},{"type":"TextBlock","text":"An alert for **endpoint-name** has been triggered due to having failed 3 time(s) in a row.","wrap":true},{"type":"FactSet","wrap":false,"facts":[{"title":"\u0026#x274C;","value":"[CONNECTED] == true"},{"title":"\u0026#x274C;","value":"[STATUS] == 200"}]}]},"contentType":"application/vnd.microsoft.card.adaptive"}],"type":"message"}
            got:
            {"attachments":[{"content":{"type":"AdaptiveCard","version":"1.4","body":[{"type":"TextBlock","text":"\u0026#x26D1; Gatus","wrap":false,"size":"Medium","weight":"Bolder"},{"type":"TextBlock","text":"An alert for **endpoint-name** has been triggered due to having failed 3 time(s) in a row.","wrap":true},{"type":"FactSet","wrap":false,"facts":[{"title":"\u0026#x274C;","value":"[CONNECTED] == true"},{"title":"\u0026#x274C;","value":"[STATUS] == 200"}]}]},"contentType":"application/vnd.microsoft.card.adaptive"}],"type":"message"}
    --- FAIL: TestAlertProvider_buildRequestBody/resolved (0.00s)
        teamsworkflows_test.go:189: expected:
            {"attachments":[{"content":{"type":"AdaptiveCard","version":"1.4","body":[{"type":"TextBlock","text":"\u0026#x1f6a8 Gatus","wrap":false,"size":"Medium","weight":"Bolder"},{"type":"TextBlock","text":"An alert for **endpoint-name** has been resolved after passing successfully 5 time(s) in a row.","wrap":true},{"type":"FactSet","wrap":false,"facts":[{"title":"\u0026#x2705;","value":"[CONNECTED] == true"},{"title":"\u0026#x2705;","value":"[STATUS] == 200"}]}]},"contentType":"application/vnd.microsoft.card.adaptive"}],"type":"message"}
            got:
            {"attachments":[{"content":{"type":"AdaptiveCard","version":"1.4","body":[{"type":"TextBlock","text":"\u0026#x26D1; Gatus","wrap":false,"size":"Medium","weight":"Bolder"},{"type":"TextBlock","text":"An alert for **endpoint-name** has been resolved after passing successfully 5 time(s) in a row.","wrap":true},{"type":"FactSet","wrap":false,"facts":[{"title":"\u0026#x2705;","value":"[CONNECTED] == true"},{"title":"\u0026#x2705;","value":"[STATUS] == 200"}]}]},"contentType":"application/vnd.microsoft.card.adaptive"}],"type":"message"}
    --- FAIL: TestAlertProvider_buildRequestBody/resolved-with-no-conditions (0.00s)
        teamsworkflows_test.go:189: expected:
            {"attachments":[{"content":{"type":"AdaptiveCard","version":"1.4","body":[{"type":"TextBlock","text":"\u0026#x1f6a8 Gatus","wrap":false,"size":"Medium","weight":"Bolder"},{"type":"TextBlock","text":"An alert for **endpoint-name** has been resolved after passing successfully 5 time(s) in a row.","wrap":true},{"type":"FactSet","wrap":false}]},"contentType":"application/vnd.microsoft.card.adaptive"}],"type":"message"}
            got:
            {"attachments":[{"content":{"type":"AdaptiveCard","version":"1.4","body":[{"type":"TextBlock","text":"\u0026#x26D1; Gatus","wrap":false,"size":"Medium","weight":"Bolder"},{"type":"TextBlock","text":"An alert for **endpoint-name** has been resolved after passing successfully 5 time(s) in a row.","wrap":true},{"type":"FactSet","wrap":false}]},"contentType":"application/vnd.microsoft.card.adaptive"}],"type":"message"}
FAIL
coverage: 93.5% of statements
FAIL	github.com/TwiN/gatus/v5/alerting/provider/teamsworkflows	0.026s
```

**Review by TwiN — CHANGES_REQUESTED** [review]
@JamesHillyard Looks like some of your tests are failing

```
 --- FAIL: TestAlertProvider_buildRequestBody (0.00s)
    --- FAIL: TestAlertProvider_buildRequestBody/triggered (0.00s)
        teamsworkflows_test.go:189: expected:
            {"attachments":[{"content":{"type":"AdaptiveCard","version":"1.4","body":[{"type":"TextBlock","text":"\u0026#x1f6a8 Gatus","wrap":false,"size":"Medium","weight":"Bolder"},{"type":"TextBlock","text":"An alert for **endpoint-name** has been triggered due to having failed 3 time(s) in a row.","wrap":true},{"type":"FactSet","wrap":false,"facts":[{"title":"\u0026#x274C;","value":"[CONNECTED] == true"},{"title":"\u0026#x274C;","value":"[STATUS] == 200"}]}]},"contentType":"application/vnd.microsoft.card.adaptive"}],"type":"message"}
            got:
            {"attachments":[{"content":{"type":"AdaptiveCard","version":"1.4","body":[{"type":"TextBlock","text":"\u0026#x26D1; Gatus","wrap":false,"size":"Medium","weight":"Bolder"},{"type":"TextBlock","text":"An alert for **endpoint-name** has been triggered due to having failed 3 time(s) in a row.","wrap":true},{"type":"FactSet","wrap":false,"facts":[{"title":"\u0026#x274C;","value":"[CONNECTED] == true"},{"title":"\u0026#x274C;","value":"[STATUS] == 200"}]}]},"contentType":"application/vnd.microsoft.card.adaptive"}],"type":"message"}
    --- FAIL: TestAlertProvider_buildRequestBody/resolved (0.00s)
        teamsworkflows_test.go:189: expected:
            {"attachments":[{"content":{"type":"AdaptiveCard","version":"1.4","body":[{"type":"TextBlock","text":"\u0026#x1f6a8 Gatus","wrap":false,"size":"Medium","weight":"Bolder"},{"type":"TextBlock","text":"An alert for **endpoint-name** has been resolved after passing successfully 5 time(s) in a row.","wrap":true},{"type":"FactSet","wrap":false,"facts":[{"title":"\u0026#x2705;","value":"[CONNECTED] == true"},{"title":"\u0026#x2705;","value":"[STATUS] == 200"}]}]},"contentType":"application/vnd.microsoft.card.adaptive"}],"type":"message"}
            got:
            {"attachments":[{"content":{"type":"AdaptiveCard","version":"1.4","body":[{"type":"TextBlock","text":"\u0026#x26D1; Gatus","wrap":false,"size":"Medium","weight":"Bolder"},{"type":"TextBlock","text":"An alert for **endpoint-name** has been resolved after passing successfully 5 time(s) in a row.","wrap":true},{"type":"FactSet","wrap":false,"facts":[{"title":"\u0026#x2705;","value":"[CONNECTED] == true"},{"title":"\u0026#x2705;","value":"[STATUS] == 200"}]}]},"contentType":"application/vnd.microsoft.card.adaptive"}],"type":"message"}
    --- FAIL: TestAlertProvider_buildRequestBody/resolved-with-no-conditions (0.00s)
        teamsworkflows_test.go:189: expected:
            {"attachments":[{"content":{"type":"AdaptiveCard","version":"1.4","body":[{"type":"TextBlock","text":"\u0026#x1f6a8 Gatus","wrap":false,"size":"Medium","weight":"Bolder"},{"type":"TextBlock","text":"An alert for **endpoint-name** has been resolved after passing successfully 5 time(s) in a row.","wrap":true},{"type":"FactSet","wrap":false}]},"contentType":"application/vnd.microsoft.card.adaptive"}],"type":"message"}
            got:
            {"attachments":[{"content":{"type":"AdaptiveCard","version":"1.4","body":[{"type":"TextBlock","text":"\u0026#x26D1; Gatus","wrap":false,"size":"Medium","weight":"Bolder"},{"type":"TextBlock","text":"An alert for **endpoint-name** has been resolved after passing successfully 5 time(s) in a row.","wrap":true},{"type":"FactSet","wrap":false}]},"contentType":"application/vnd.microsoft.card.adaptive"}],"type":"message"}
FAIL
coverage: 93.5% of statements
FAIL	github.com/TwiN/gatus/v5/alerting/provider/teamsworkflows	0.026s
```

**Comment by JamesHillyard** [comment]
@TwiN Looks like the text was expecting a broken teams emoji.. I must have changed it last minute and not updated the test. I've now fixed it and running `go test -v .\alerting\provider\teamsworkflows\` results in all tests passing.

**Comment by deebify** [comment]
When can I use the new teams-workflows workflow?

**Review by TwiN — APPROVED** [review]

**Comment by TwiN** [comment]
@JamesHillyard Great work & thank you for the contribution!

**Comment by TwiN** [comment]
@liamjones-pw I usually let new features soak in latest for some time to give people time to report issues. If people are already using the feature and haven't reported any issues, I can create a release for it this weekend 👀

**Comment by TwiN** [comment]
@liamjones-pw I just released v5.13.0

**Pull request merged**